packages/network-controller/src/NetworkController.ts

```diff
@@ -1353,6 +1953,458 @@ export class NetworkController extends BaseController<
+    autoManagedNetworkClientRegistry,
+  }: {
+    networkClientOperations: NetworkClientOperation[];
+    autoManagedNetworkClientRegistry: AutoManagedNetworkClientRegistry;
+  }) {
+    const removedRpcEndpoints = networkClientOperations
+      .filter(
+        (
+          networkClientOperation,
+        ): networkClientOperation is RemoveNetworkClientOperation => {
+          return networkClientOperation.type === 'remove';
+        },
+      )
+      .map((networkClientOperation) => networkClientOperation.rpcEndpoint)
+      .concat(
+        networkClientOperations
+          .filter(
+            (
+              networkClientOperation,
+            ): networkClientOperation is ReplaceNetworkClientOperation => {
+              return networkClientOperation.type === 'replace';
+            },
+          )
+          .map(
+            (networkClientOperation) => networkClientOperation.oldRpcEndpoint,
+          ),
+      );
+
+    for (const rpcEndpoint of removedRpcEndpoints) {
+      const networkClient = this.getNetworkClientById(
+        rpcEndpoint.networkClientId,
+      );
+      networkClient.destroy();
+      delete autoManagedNetworkClientRegistry[networkClient.configuration.type][
+        rpcEndpoint.networkClientId
+      ];
+    }
+  }
+
+  /**
+   * Updates `networkConfigurationsByChainId` in state depending on whether a
+   * network is being added, updated, or removed.
+   *
+   * - The existing network configuration will be removed when a network is
+   * being filed under a different chain or removed.
+   * - A network configuration will be stored when a network is being added or
+   * when a network is being updated.
+   *
+   * @param args - The arguments to this function.
+   */
+  #updateNetworkConfigurations(
+    args: { state: Draft<NetworkState> } & (
+      | {
+          mode: 'add';
+          networkFields: AddNetworkFields;
+          networkConfigurationToPersist: NetworkConfiguration;
+        }
+      | {
+          mode: 'update';
+          networkFields: UpdateNetworkFields;
+          networkConfigurationToPersist: NetworkConfiguration;
+          existingNetworkConfiguration: NetworkConfiguration;
+        }
+      | {
+          mode: 'remove';
+          existingNetworkConfiguration: NetworkConfiguration;
+        }
+    ),
+  ) {
+    const { state, mode } = args;
+
+    if (
+      mode === 'remove' ||
+      (mode === 'update' &&
+        args.networkFields.chainId !==
+          args.existingNetworkConfiguration.chainId)
+    ) {
+      delete state.networkConfigurationsByChainId[
+        args.existingNetworkConfiguration.chainId
+      ];
+    }
+
+    if (mode === 'add' || mode === 'update') {
+      state.networkConfigurationsByChainId[args.networkFields.chainId] =
+        args.networkConfigurationToPersist;
+    }
+
+    this.#networkConfigurationsByNetworkClientId =
+      buildNetworkConfigurationsByNetworkClientId(
+        this.state.networkConfigurationsByChainId,
```

Comment:
```suggestion
        state.networkConfigurationsByChainId,
```

I think this fixes a sneaky bug in the context of adding or updating a network.

The network is added to `state.networkConfigurationsByChainId` a few lines above.  But it's not yet added to `this.state.networkConfigurationsByChainId` at this point, because we're still inside an update function.  Without this fix, the new network does not get passed to `buildNetworkConfigurationsByNetworkClientId`.

![Screenshot 2024-08-26 at 8 35 38 PM](https://github.com/user-attachments/assets/6248a170-916b-407d-9cdd-0982e5386b6c)

Reply:
I committed this change so I can publish a preview version containing it

Reply:
Had to take another pass and fix it this way instead: https://github.com/MetaMask/core/pull/4286/commits/02e7ff99c640127d9b08c80f0951f1141c16b3c6